File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.go
package main

import "fmt"

func longestPalindrome(s string) string {
	n := len(s)
	if n < 2 {
		return s
	}

	start, maxLength := 0, 1

	for i := 0; i < n-1; i++ {
		for j := i + 1; j < n; j++ {
			if isPalindrome(s, i, j) && j-i+1 > maxLength {
				start = i
				maxLength = j - i + 1
			}
		}
	}

	return s[start : start+maxLength]
}

func isPalindrome(s string, left, right int) bool {
	for left < right {
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}
	return true
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    if (s.length < 2) return s
    let start = 0, maxLength = 1
    
    const expandAroundCenter = (left, right) => {
        while (s[left] === s[right] && left >= 0 && right < s.length) {
            if (right - left + 1 > maxLength) {
                start = left
                maxLength = right - left + 1
            }
            
            left -= 1
            right += 1
        }
    }
    
    for (let i = 0; i < s.length; i++) {
        expandAroundCenter(i - 1, i + 1)
        expandAroundCenter(i, i + 1)
    }
    
    return s.substring(start, start + maxLength)
}
